File: lib/analyzer.py
from __future__ import annotations

from dataclasses import dataclass

import pandas as pd


@dataclass
class CategorySummary:
    category: str
    total: float
    count: int
    pct_of_income: float
    pct_of_expenses: float | None


@dataclass
class MerchantSummary:
    merchant_name: str
    category: str
    total: float


@dataclass
class AnalysisResult:
    total_income: float
    total_expenses: float
    net_savings: float
    savings_rate: float
    income_categories: list[CategorySummary]
    expense_categories: list[CategorySummary]
    top_merchants: list[MerchantSummary]
    unusual_expenses: list[CategorySummary]
    insights: list[str]


def _category_totals(df: pd.DataFrame, transaction_type: str) -> pd.DataFrame:
    subset = df[df["transaction_type"] == transaction_type]
    grouped = (
        subset.groupby("category", as_index=False)["abs_amount"]
        .agg(total="sum", count="count")
        .sort_values("total", ascending=False)
    )
    return grouped


def analyze_transactions(df: pd.DataFrame) -> AnalysisResult:
    total_income = float(df.loc[df["transaction_type"] == "income", "abs_amount"].sum())
    total_expenses = float(df.loc[df["transaction_type"] == "expense", "abs_amount"].sum())
    net_savings = total_income - total_expenses
    savings_rate = (net_savings / total_income * 100) if total_income else 0.0

    income_df = _category_totals(df, "income")
    expense_df = _category_totals(df, "expense")

    income_categories = [
        CategorySummary(
            category=row.category,
            total=float(row.total),
            count=int(row.count),
            pct_of_income=(row.total / total_income * 100) if total_income else 0.0,
            pct_of_expenses=None,
        )
        for row in income_df.itertuples(index=False)
    ]

    expense_categories = [
        CategorySummary(
            category=row.category,
            total=float(row.total),
            count=int(row.count),
            pct_of_income=(row.total / total_income * 100) if total_income else 0.0,
            pct_of_expenses=(row.total / total_expenses * 100) if total_expenses else 0.0,
        )
        for row in expense_df.itertuples(index=False)
    ]

    merchant_df = (
        df[df["transaction_type"] == "expense"]
        .groupby(["merchant_name", "category"], as_index=False)["abs_amount"]
        .sum()
        .sort_values("abs_amount", ascending=False)
        .head(10)
    )
    top_merchants = [
        MerchantSummary(
            merchant_name=row.merchant_name,
            category=row.category,
            total=float(row.abs_amount),
        )
        for row in merchant_df.itertuples(index=False)
    ]

    unusual_expenses = _detect_unusual_expenses(expense_categories)
    insights = _build_insights(
        total_income=total_income,
        total_expenses=total_expenses,
        net_savings=net_savings,
        savings_rate=savings_rate,
        expense_categories=expense_categories,
        unusual_expenses=unusual_expenses,
    )

    return AnalysisResult(
        total_income=total_income,
        total_expenses=total_expenses,
        net_savings=net_savings,
        savings_rate=savings_rate,
        income_categories=income_categories,
        expense_categories=expense_categories,
        top_merchants=top_merchants,
        unusual_expenses=unusual_expenses,
        insights=insights,
    )


def _detect_unusual_expenses(expense_categories: list[CategorySummary]) -> list[CategorySummary]:
    if len(expense_categories) < 2:
        return []

    totals = pd.Series([item.total for item in expense_categories])
    mean_value = totals.mean()
    std_value = totals.std(ddof=0)
    threshold = max(mean_value * 1.5, mean_value + std_value)

    return [item for item in expense_categories if item.total >= threshold]


def _build_insights(
    total_income: float,
    total_expenses: float,
    net_savings: float,
    savings_rate: float,
    expense_categories: list[CategorySummary],
    unusual_expenses: list[CategorySummary],
) -> list[str]:
    insights: list[str] = []

    if total_income == 0:
        insights.append("No income categories were found. Add rows with categories like Salary or Pension.")
    if total_expenses == 0:
        insights.append("No expense categories were found.")

    if total_income and total_expenses:
        if savings_rate >= 20:
            insights.append(f"You are saving {savings_rate:.1f}% of income, which is a healthy rate.")
        elif savings_rate >= 0:
            insights.append(f"You are saving {savings_rate:.1f}% of income. Consider targeting 20% or more.")
        else:
            insights.append(
                f"Expenses exceed income by ${abs(net_savings):,.2f}. Review your largest spending categories."
            )

    if expense_categories:
        largest = expense_categories[0]
        insights.append(
            f"Largest expense category is {largest.category} at ${largest.total:,.2f} "
            f"({largest.pct_of_expenses:.1f}% of expenses, {largest.pct_of_income:.1f}% of income)."
        )

    for item in unusual_expenses[:3]:
        insights.append(
            f"{item.category} looks unusually high at ${item.total:,.2f} "
            f"({item.pct_of_expenses:.1f}% of total expenses)."
        )

    return insights

File: frontend/lib/analyzer.ts
import {
  filterExpenseTransactions,
  filterIncomeTransactions,
  isExpenseTransaction,
} from "@/lib/transaction-filters";
import { CategorySummary, MerchantSummary, PeriodAnalysis, Transaction } from "@/lib/types";
import { round2 } from "@/lib/utils";

function groupExpenseCategories(rows: Transaction[]) {
  const map = new Map<string, { total: number; count: number }>();
  for (const row of filterExpenseTransactions(rows)) {
    const current = map.get(row.category) ?? { total: 0, count: 0 };
    current.total += row.abs_amount;
    current.count += 1;
    map.set(row.category, current);
  }
  return [...map.entries()]
    .map(([category, value]) => ({ category, ...value }))
    .sort((a, b) => b.total - a.total);
}

function groupIncomeCategories(rows: Transaction[]) {
  const map = new Map<string, { total: number; count: number }>();
  for (const row of filterIncomeTransactions(rows)) {
    const current = map.get(row.category) ?? { total: 0, count: 0 };
    current.total += row.abs_amount;
    current.count += 1;
    map.set(row.category, current);
  }
  return [...map.entries()]
    .map(([category, value]) => ({ category, ...value }))
    .sort((a, b) => b.total - a.total);
}

function detectUnusual(expenseCategories: CategorySummary[]) {
  if (expenseCategories.length < 2) return [];
  const totals = expenseCategories.map((item) => item.total);
  const avg = totals.reduce((sum, value) => sum + value, 0) / totals.length;
  const variance =
    totals.reduce((sum, value) => sum + (value - avg) ** 2, 0) / totals.length;
  const std = Math.sqrt(variance);
  const threshold = Math.max(avg * 1.5, avg + std);
  return expenseCategories.filter((item) => item.total >= threshold);
}

export function analyzeTransactions(rows: Transaction[]): PeriodAnalysis {
  const incomeRows = filterIncomeTransactions(rows);
  const expenseRows = filterExpenseTransactions(rows);
  const total_income = round2(incomeRows.reduce((sum, row) => sum + row.abs_amount, 0));
  const expenseGroups = groupExpenseCategories(rows);
  const total_expenses = round2(expenseGroups.reduce((sum, item) => sum + item.total, 0));
  const net_savings = round2(total_income - total_expenses);
  const savings_rate = total_income ? round2((net_savings / total_income) * 100) : 0;

  const income_categories: CategorySummary[] = groupIncomeCategories(rows).map((item) => ({
    category: item.category,
    total: round2(item.total),
    count: item.count,
    pct_of_income: total_income ? round2((item.total / total_income) * 100) : 0,
    pct_of_expenses: null,
  }));

  const expense_categories: CategorySummary[] = expenseGroups.map((item) => ({
    category: item.category,
    total: round2(item.total),
    count: item.count,
    pct_of_income: total_income ? round2((item.total / total_income) * 100) : 0,
    pct_of_expenses: total_expenses ? round2((item.total / total_expenses) * 100) : 0,
  }));

  const merchantMap = new Map<string, MerchantSummary>();
  for (const row of expenseRows) {
    const key = `${row.merchant_name}::${row.category}`;
    const current = merchantMap.get(key) ?? {
      merchant_name: row.merchant_name,
      category: row.category,
      total: 0,
    };
    current.total += row.abs_amount;
    merchantMap.set(key, current);
  }
  const top_merchants = [...merchantMap.values()]
    .sort((a, b) => b.total - a.total)
    .slice(0, 10)
    .map((item) => ({ ...item, total: round2(item.total) }));

  const unusual_expenses = detectUnusual(expense_categories);
  const insights: string[] = [];
  if (!total_income) {
    insights.push("No income categories were found. Add rows with categories like Salary or Pension.");
  }
  if (!total_expenses) insights.push("No expense categories were found.");
  if (total_income && total_expenses) {
    if (savings_rate >= 20) {
      insights.push(`You are saving ${savings_rate.toFixed(1)}% of income, which is a healthy rate.`);
    } else if (savings_rate >= 0) {
      insights.push(`You are saving ${savings_rate.toFixed(1)}% of income. Consider targeting 20% or more.`);
    } else {
      insights.push(
        `Expenses exceed income by $${Math.abs(net_savings).toFixed(2)}. Review your largest spending categories.`
      );
    }
  }
  if (expense_categories[0]) {
    const largest = expense_categories[0];
    insights.push(
      `Largest expense category is ${largest.category} at $${largest.total.toFixed(2)} (${largest.pct_of_expenses?.toFixed(1)}% of expenses, ${largest.pct_of_income.toFixed(1)}% of income).`
    );
  }
  for (const item of unusual_expenses.slice(0, 3)) {
    insights.push(
      `${item.category} looks unusually high at $${item.total.toFixed(2)} (${item.pct_of_expenses?.toFixed(1)}% of total expenses).`
    );
  }

  return {
    total_income,
    total_expenses,
    net_savings,
    savings_rate,
    income_categories,
    expense_categories,
    top_merchants,
    unusual_expenses,
    insights,
  };
}

export function expenseCategoryTotal(rows: Transaction[], category: string) {
  return round2(
    rows
      .filter((row) => isExpenseTransaction(row) && row.category === category)
      .reduce((sum, row) => sum + row.abs_amount, 0)
  );
}
